junit-platform-engine/src/main/java/org/junit/platform/engine/TestDescriptor.java

```diff
@@ -172,6 +176,30 @@ default Set<? extends TestDescriptor> getDescendants() {
 	 */
 	void removeFromHierarchy();
 
+	/**
+	 * Order all children from this descriptor.
+	 *
+	 * <p>The {@code orderer} is provided a modifiable list of children in
+	 * this test descriptor. Never {@code null}. The {@code orderer} may return
+	 * the elements of list in any order but may not add or remove descriptors.
+	 * <p>
+	 * If descriptors were added or removed this method must throw a
+	 * {@link org.junit.platform.commons.JUnitException JUnitException} explaining
+	 * that an {@code orderer} may not remove children from this descriptor.
+	 *
+	 * @param orderer a unary operator to order the children from this test descriptor.
+	 */
+	@API(since = "5.12", status = EXPERIMENTAL)
+	default void orderChildren(UnaryOperator<List<TestDescriptor>> orderer) {
+		Preconditions.notNull(orderer, "orderer must not be null");
+		Set<? extends TestDescriptor> children = getChildren();
+		Set<? extends TestDescriptor> orderedChildren = new LinkedHashSet<>(orderer.apply(new ArrayList<>(children)));
+		boolean unmodified = children.equals(orderedChildren);
+		Preconditions.condition(unmodified, "orderer may not add or remove test descriptors");
```

Comment:
I think we're lacking a test where a misbehaving orderer that removes a descriptor still gets its ordering for the remaining descriptors applied. The check introduced by this line will cause an exception to be thrown in this case which gets caught and logged in `AbstractOrderingVisitor.doWithMatchingDescriptor`.

Could you please add such a test (or extend an existing one) and make sure to include the removed descriptors in the `UnaryOperator` implementation to restore the original behavior?

Reply:
I think the logical conclusion of this request is the remove this check and use the order provided by the orderer as a suggestion for the actual order.

Reply:
I've pushed a commit that reinstates the condition because I think it's better to be strict here. @mpkorstanje Could you please take a look?

Reply:
Ah. I see what you meant.

Left a few comments.